### ISSUE TITLE
Infura IPFS SSL certificate verification fails

### ISSUE THREAD

**Issue by njgheorghita** [ipfs/py-ipfs-api#184]
Hello!

I just updated to `ipfshttpclient` from `py-ipfs-api` and now all Infura IPFS requests are failing.

Infura IPFS multiaddr:
`self.base_uri = "/dns/ipfs.infura.io/tcp/5001/https"` 
(I've also tried with `/dns4/`)

Connecting to Infura IPFS
```python
self.client = ipfshttpclient.connect(self.base_uri)
```

Stack trace: 
```sh
ethpm/backends/ipfs.py:60: in __init__
    self.client = ipfshttpclient.connect(self.base_uri)
venv/lib/python3.6/site-packages/ipfshttpclient/client/__init__.py:102: in connect
    assert_version(client.version()['Version'])
venv/lib/python3.6/site-packages/ipfshttpclient/client/base.py:15: in wrapper
    result = func(*args, **kwargs)
venv/lib/python3.6/site-packages/ipfshttpclient/client/miscellaneous.py:204: in version
    return self._client.request('/version', decoder='json', **kwargs)
venv/lib/python3.6/site-packages/ipfshttpclient/http.py:44: in wrapper
    return func(self, *args, **merged)
venv/lib/python3.6/site-packages/ipfshttpclient/http.py:364: in request
    files, headers, data, timeout=timeout)
venv/lib/python3.6/site-packages/ipfshttpclient/http.py:282: in _request
    timeout=timeout)
venv/lib/python3.6/site-packages/ipfshttpclient/http.py:249: in _do_request
    six.raise_from(exceptions.ConnectionError(error), error)
<string>:3: in raise_from
    ???
E   ipfshttpclient.exceptions.ConnectionError: SSLError: HTTPSConnectionPool(host='ipfs.infura.io', port=5001): Max retries exceeded with url: /api/v0/version?stream-channels=true (Caused by SSLError(SSLError("bad handshake: Error([('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')],)",),))
```

I'm not sure what's the cause for the SSL certificate verification failure. 
There doesn't seem to be any problem with [infura's ssl certificates](https://www.sslshopper.com/ssl-checker.html#hostname=ipfs.infura.io). 
It appears that the multiaddr is being correctly parsed, since the request is being made to https://ipfs.infura.io:5001/api/v0/version. 

I've tried adding Infura's CA to `certifi`'s `cacert.pem` without any luck. 
I've also tried hard-coding `requests.request`'s `verify` parameter to `False` just to see if it works, but no luck.

Any thoughts or ideas on get these requests working again would be great, cheers!

**Comment by alexander255** [comment]
Thanks for reporting this! Unfortunately it appears you already tested most of the things I may have thought of. :slightly_smiling_face: 

Also, I just tested running `python3 -c "import ipfshttpclient; ipfshttpclient.connect('/dns/ipfs.infura.io/tcp/5001/https')"` and I see no errors unfortunately.
Does running `curl -L https://ipfs.infura.io:5001/api/v0/version` work for you?

If that works please try editing the file `ipfshttpclient/http.py` and replace:

  * At top of the file replace the line: `from . import requests_wrapper as requests` with `import requests`
  * In `HTTPClient.__init__` drop the line: `"family": family` within the declaration of `self._kwargs`

Then test again. Thanks!

**Comment by njgheorghita** [comment]
Thanks for the quick reply! Interestingly for me, `python3 -c "import ipfshttpclient; ipfshttpclient.connect('/dns/ipfs.infura.io/tcp/5001/https')"` still reports the same error, but there's no error for `curl -L https://ipfs.infura.io:5001/api/v0/version`. 

But, after I made those changes, everything's working great!

**Comment by njgheorghita** [comment]
Here's my output from `pip freeze` just in case it's helpful. 
```
alabaster==0.7.12
apipkg==1.5
appdirs==1.4.3
appnope==0.1.0
argh==0.26.2
asn1crypto==0.24.0
atomicwrites==1.3.0
attrdict==2.0.1
attrs==19.1.0
Babel==2.6.0
backcall==0.1.0
base58==1.0.3
black==19.3b0
bleach==3.1.0
bumpversion==0.5.3
certifi==2019.3.9
cffi==1.12.3
chardet==3.0.4
Click==7.0
colorama==0.4.1
cryptography==2.6.1
cytoolz==0.9.0.1
decorator==4.4.0
docopt==0.6.2
docutils==0.14
entrypoints==0.3
eth-abi==2.0.0b9
eth-account==0.4.0
eth-bloom==1.0.3
eth-hash==0.2.0
eth-keyfile==0.5.1
eth-keys==0.2.2
eth-rlp==0.1.2
eth-tester==0.1.0b39
eth-typing==2.1.0
eth-utils==1.6.0
ethpm==0.1.4a16
-e git+https://github.com/ethpm/ethpm-cli.git@0f5d5dcccff57ad020d3a9381aca4755eafcd173#egg=ethpm_cli
execnet==1.6.0
flake8==3.7.7
hexbytes==0.1.0
idna==2.8
imagesize==1.1.0
ipfshttpclient==0.4.11
ipython==7.5.0
ipython-genutils==0.2.0
isort==4.3.20
jedi==0.13.3
Jinja2==2.10.1
jsonschema==2.6.0
lru-dict==1.1.6
MarkupSafe==1.1.1
mccabe==0.6.1
more-itertools==7.0.0
multiaddr==0.0.7
mypy==0.701
mypy-extensions==0.4.1
netaddr==0.7.19
packaging==19.0
parsimonious==0.8.1
parso==0.4.0
pathtools==0.1.2
pexpect==4.7.0
pickleshare==0.7.5
pkginfo==1.5.0.1
pluggy==0.11.0
prompt-toolkit==2.0.9
protobuf==3.7.1
ptyprocess==0.6.0
py==1.8.0
py-ecc==1.6.0
py-evm==0.2.0a42
py-geth==2.1.0
pycodestyle==2.5.0
pycparser==2.19
pycryptodome==3.8.1
pydocstyle==3.0.0
pyethash==0.1.27
pyflakes==2.1.1
Pygments==2.4.0
pyOpenSSL==19.0.0
pyparsing==2.4.0
pysha3==1.0.2
pytest==4.5.0
pytest-ethereum==0.1.3a6
pytest-forked==1.0.2
pytest-watch==4.2.0
pytest-xdist==1.28.0
pytz==2019.1
PyYAML==5.1
readme-renderer==24.0
requests==2.22.0
requests-toolbelt==0.9.1
rlp==1.1.0
semantic-version==2.6.0
six==1.12.0
snowballstemmer==1.2.1
Sphinx==1.8.5
sphinx-rtd-theme==0.4.3
sphinxcontrib-websupport==1.1.2
toml==0.10.0
toolz==0.9.0
tox==2.9.1
tqdm==4.32.1
traitlets==4.3.2
trie==1.4.0
twine==1.13.0
typed-ast==1.3.5
urllib3==1.25.2
varint==1.0.2
virtualenv==16.6.0
watchdog==0.9.0
wcwidth==0.1.7
web3==5.0.0b2
webencodings==0.5.1
websockets==7.0
```

**Comment by alexander255** [comment]
Thanks! In the meantime I found that looks like an incompatibility with `urllib3==1.25.*`, but it's good to know that the suggested quick-fix works. I'll try to push a mitigation for this (either the suggested fix or something else) this evening.

**Comment by njgheorghita** [comment]
Awesome, thanks so much @alexander255 !

**Comment by alexander255** [comment]
Fix is in the pipeline: #185

**Comment by alexander255** [comment]
I have pushed version 0.4.12 just now that should resolve this issue. Thanks for reporting this!
